core/src/mast/debuginfo/mod.rs

```diff
@@ -1,7 +1,130 @@
+//!
+//! The debug info is always available at the `MastForest` level (as per issue #1821),
+//! but may be conditionally included during assembly to maintain backward compatibility.
+//! Decorators are only executed when the processor is running in debug mode, allowing
+//! debug information to be available for debugging and error reporting without
+//! impacting performance in production execution.
+//!
+//! # Debug Mode Semantics
+//!
+//! Debug mode is controlled via [`ExecutionOptions`](air::options::ExecutionOptions):
+//! - `with_debugging(true)` enables debug mode explicitly
+//! - `with_tracing()` automatically enables debug mode (tracing requires debug info)
+//! - By default, debug mode is disabled for maximum performance
+//!
+//! When debug mode is disabled:
+//! - Debug decorators are not executed
+//! - Trace decorators are not executed
+//! - Assembly operation decorators are not recorded
+//! - before_enter/after_exit decorators are not executed
+//!
+//! When debug mode is enabled:
+//! - All decorator types are executed according to their semantics
+//! - Debug decorators trigger host callbacks for breakpoints
+//! - Trace decorators trigger host callbacks for tracing
+//! - Assembly operation decorators provide source mapping information
+//! - before_enter/after_exit decorators execute around node execution
+//!
+//! # Production Builds
+//!
+//! The `DebugInfo` can be stripped for production builds using the [`strip()`](Self::strip)
+//! method, which removes decorators while preserving critical information. This allows
+//! backward compatibility while enabling size optimization for deployment.
+
+use alloc::{collections::BTreeMap, sync::Arc};
+
+use miden_utils_indexing::IndexVec;
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+use super::{Decorator, DecoratorId};
+
 mod decorator_storage;
 pub use decorator_storage::{
     DecoratedLinks, DecoratedLinksIter, DecoratorIndexError, OpToDecoratorIds,
 };
 
 mod node_decorator_storage;
 pub use node_decorator_storage::NodeToDecoratorIds;
+
+// DEBUG INFO
+// ================================================================================================
+
+/// Debug information for a MAST forest, containing decorators and error messages.
+/// This is always present in a MastForest (as per issue #1821), but may be "stripped"
+/// in the future for release builds.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct DebugInfo {
+    /// All decorators in the MAST forest.
+    pub decorators: IndexVec<DecoratorId, Decorator>,
+
+    /// Efficient access to decorators per operation per node.
+    pub op_decorator_storage: OpToDecoratorIds,
+
+    /// Efficient storage for node-level decorators (before_enter and after_exit).
+    pub node_decorator_storage: NodeToDecoratorIds,
+
+    /// Maps error codes to error messages.
+    pub error_codes: BTreeMap<u64, Arc<str>>,
+}
```

Comment:
Would it not be better to make the fields private to hide the internal implementation details of this struct? We could just expose the relevant methods from this struct (e.g., `decorators_for_op()`, `get_decorator_by_id()` etc.).